### PR TITLE
Feature reimbursement forms

### PR DESCRIPTION
**Pull request by KaiKaiai**

### CODE REVIEW

File: src/auth/google.ts [hunk 1]
```diff
@@ -17,31 +17,28 @@ provider.addScope("https://www.googleapis.com/auth/drive.file");
...
-        throw new Error("Please login using CISSA account");
-      }
-    })
-    .catch((error) => {
-      onError(error);
-    });
+export async function googleSignIn(): Promise<User> {
+  const result: UserCredential = await signInWithPopup(auth, provider);
+  const credential = GoogleAuthProvider.credentialFromResult(result);
+  console.log(credential);
```
Comment (thebowenfeng) [inline src/auth/google.ts]: Prolly shouldn't be logging this lol

File: src/auth/types.ts [hunk 2]
```diff
@@ -4,8 +4,34 @@ export interface User {
...
+  bsb?: string;
+  accountNum?: string;
+  // TODO: add an account name
 }
+
+// creates an user instance
+export function createUser(
+  obj: any
+): User {
+  if (obj == null){
```
Comment (thebowenfeng) [inline src/auth/types.ts:20]: Might not be a big issue depending on how your logic is setup but shallow equality here means undefined obj also passes the check (i.e when obj is undefined).

File: src/database/reimbursement.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import {app} from "../config";
+import {getFirestore, doc, setDoc, getDoc, getDocs, collection, query, where, orderBy} from "firebase/firestore";
+import {createUser, User} from "../auth/types";
+
+const db = getFirestore(app);
+
+export type DepartmentEnum = "IT" | "Events" | "Competition" | "Education" | "Industry" | "Project" | "Diversity" | "Publicity" | "Product";
```
Comment (thebowenfeng) [inline src/database/reimbursement.ts:7]: Easier if you define an actual enum and the can just access the property like Department.Publicity

File: src/database/reimbursement.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  const user = createUser(userData);
+
+  // if lacking banking information
+  if (user.accountNum == null || user.bsb == null) {
+    throw new Error("no banking information found");
+  }
+
+  // store the document
+  const ref = doc(collection(db, "reimbursement"));
+  await setDoc(ref, submission);
```
Comment (thebowenfeng) [inline src/database/reimbursement.ts]: Should prolly handle errors at this point in case setDoc fails.

File: src/database/reimbursement.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+export async function getMyReimbursement(user: User): Promise<Reimbursement[]> {
+  const q = query(
+    collection(db, "reimbursement"),
+    where("state", "==", "Active"),
+    where("userid", "==", user.id),
+    orderBy("purchaseDate", "desc")
+  );
+
+  const snapshot = await getDocs(q);
+  let result: Reimbursement[] = [];
```
Comment (thebowenfeng) [inline src/database/reimbursement.ts]: Can be replaced by map

File: src/database/reimbursement.ts (hunk 3 above)
Comment (khangluong2004) [reply src/database/reimbursement.ts:7]: I kept it as a group of strings so it's easier to use those strings as selection options directly though

File: src/auth/types.ts (hunk 2 above)
Comment (Troppydash) [reply src/auth/types.ts:20]: Probably keep it as shallow compare; we need undefined to be truthy else the destructuring below fails.

**Pull request merged**